src/iommufd/iommufd.c

```diff
@@ -0,0 +1,281 @@
+		err = errno;
+		log_debug("opendir failed (errno %d)\n", err);
+		goto out;
+	}
+
+	do {
+		dentry = readdir(dp);
+		if (!dentry) {
+			err = errno;
+			log_debug("readdir failed (errno %d)\n", err);
+			goto close_dir;
+		}
+
+		if (strncmp("vfio", dentry->d_name, 4) == 0)
+			break;
+	} while (dentry != NULL);
+
+	if (dentry == NULL) {
+		log_debug("Directory vfioX was not found in %s\n", path);
+		err = -1;
+		goto close_dir;
+	}
+
+	dir_nptr = dentry->d_name + 4;
+	*id = strtoul(dir_nptr, &dir_endptr, 10);
+	if (*id == ULONG_MAX || dir_nptr == dir_endptr) {
+		err = -1;
+		log_debug("Could not extrace vfio id from directory\n");
+		goto close_dir;
+	}
+
+	err = 0;
+
+close_dir:
+	if (closedir(dp)) {
+		err = errno;
+		log_debug("closedir filed (errno %d)\n", err);
+	}
+out:
+	free(path);
+	return err;
+}
+
+static int __vfio_iommufd_get_dev_fd(const char *bdf)
+{
+	unsigned long vfio_id;
+	int devfd;
+	__autofree char *devpath;
+
+	if (vfio_iommufd_get_device_id(bdf, &vfio_id)) {
+		log_error("Could not determine the vfio device id for %s\n", bdf);
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (asprintf(&devpath, "/dev/vfio/devices/vfio%ld", vfio_id) < 0) {
+		log_debug("asprintf failed\n");
+		errno = EINVAL;
+		return -1;
+	}
+
+	devfd = open(devpath, O_RDWR);
+	if (devfd < 0)
+		log_error("Error opening %s\n", devpath);
+
+	return devfd;
+}
+
+int vfio_get_device_fd(struct vfio_container *vfio, const char *bdf)
+{
+	int devfd, saved_errno;
+	struct vfio_device_bind_iommufd bind = {
+		.argsz = sizeof(bind),
+		.flags = 0
+	};
+
+	struct iommu_ioas_alloc alloc_data = {
+		.size = sizeof(alloc_data),
+		.flags = 0,
+	};
+
+	struct vfio_device_attach_iommufd_pt attach_data = {
+		.argsz = sizeof(attach_data),
+		.flags = 0,
+	};
+
+	devfd = __vfio_iommufd_get_dev_fd(bdf);
+	if (devfd < 0) {
+		saved_errno = errno;
+		log_error("failed to get the cdev file descriptor %s\n", strerror(saved_errno));
+		return devfd;
+	}
+
+	bind.iommufd = vfio->fd;
+	if (ioctl(devfd, VFIO_DEVICE_BIND_IOMMUFD, &bind)) {
+		saved_errno = errno;
+		log_error("Error binding iommufd %d to cdev %d, %s\n",
+			  vfio->fd, devfd, strerror(saved_errno));
+		goto close_dev;
+	}
+
+	if (ioctl(vfio->fd, IOMMU_IOAS_ALLOC, &alloc_data)) {
+		saved_errno = errno;
+		log_error("Error allocating IO address space %s\n", strerror(saved_errno));
+		goto close_dev;
+	}
+
+	attach_data.pt_id = alloc_data.out_ioas_id;
+	if (ioctl(devfd, VFIO_DEVICE_ATTACH_IOMMUFD_PT, &attach_data)) {
+		saved_errno = errno;
+		log_error("Error attaching cdev %d to ioas id %d, %s\n",
+			  devfd, attach_data.pt_id, strerror(saved_errno));
+		//JAG: Should we deallocate this explicitly?
+		goto close_dev;
+	}
+	vfio->ioas_id = alloc_data.out_ioas_id;
+
+	if (vfio_configure_iommu(vfio)) {
+		saved_errno = errno;
+		log_error("failed to configure iommu: %s\n", strerror(saved_errno));
+		goto close_dev;
+	}
+
+	return devfd;
+
+close_dev:
+	// closing devfd will automatically unbind it from iommufd
```

Comment:
Please use `/*` style comments.

Reply:
I was sure that checkpatch would complain about this, but it doesnt...

Reply:
done